db/fetcher/indexer_iterators.go

```diff
@@ -501,38 +431,176 @@ func createIndexIterator(
+	res := indexDesc.Unique && len(conditions) == len(indexDesc.Fields)
+	for i := 1; i < len(conditions); i++ {
+		res = res && conditions[i].op == opEq
+	}
+	return res
+}
+
+func getFieldsBytes(conditions []fieldFilterCond) ([][]byte, error) {
+	result := make([][]byte, 0, len(conditions))
+	for i := range conditions {
+		fieldVal := client.NewFieldValue(client.LWW_REGISTER, conditions[i].val)
+		keyFieldBytes, err := fieldVal.Bytes()
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, keyFieldBytes)
+	}
+	return result, nil
+}
+
+func (f *IndexFetcher) createIndexIterator() (indexIterator, error) {
+	fieldConditions := f.determineFieldFilterConditions()
+	indexDataStoreKey := core.IndexDataStoreKey{CollectionID: f.col.ID(), IndexID: f.indexDesc.ID}
+
+	matchers, err := createValueMatchers(fieldConditions)
+	if err != nil {
+		return nil, err
+	}
+
+	switch fieldConditions[0].op {
+	case opEq:
+		if isUniqueFetchByFullKey(&f.indexDesc, fieldConditions) {
+			keyFieldsBytes, err := getFieldsBytes(fieldConditions)
+			if err != nil {
+				return nil, err
 			}
-			return newInIndexIterator(iter, valArr), nil
+			return &eqSingleIndexIterator{
+				indexKey:       indexDataStoreKey,
+				keyFieldValues: keyFieldsBytes,
+				execInfo:       &f.execInfo,
+			}, nil
 		} else {
-			return &scanningIndexIterator{
-				indexKey: indexDataStoreKey,
-				matcher:  newNinIndexCmp(valArr, false),
-				execInfo: execInfo,
+			fieldVal := client.NewFieldValue(client.LWW_REGISTER, fieldConditions[0].val)
+
+			keyValueBytes, err := fieldVal.Bytes()
+			if err != nil {
+				return nil, err
+			}
+
+			// iterators for _eq filter already iterate over keys with first field value
+			// matching the filter value, so we can skip the first matcher
+			if len(matchers) > 1 {
+				matchers[0] = &anyMatcher{}
+			}
+
+			return &eqPrefixIndexIterator{
+				indexKey:      indexDataStoreKey,
+				keyFieldValue: keyValueBytes,
+				execInfo:      &f.execInfo,
+				matchers:      matchers,
 			}, nil
 		}
-	case opLike:
-		return &scanningIndexIterator{
-			indexKey: indexDataStoreKey,
-			matcher:  newLikeIndexCmp(filterVal.(string), true),
-			execInfo: execInfo,
+	case opIn:
+		inArr, ok := fieldConditions[0].val.([]any)
+		if !ok {
+			return nil, errors.New("invalid _in/_nin value")
+		}
+		keyFieldArr := make([][]byte, 0, len(inArr))
+		for _, v := range inArr {
+			fieldVal := client.NewFieldValue(client.LWW_REGISTER, v)
+			keyFieldBytes, err := fieldVal.Bytes()
+			if err != nil {
+				return nil, err
+			}
+			keyFieldArr = append(keyFieldArr, keyFieldBytes)
+		}
+
+		// iterators for _in filter already iterate over keys with first field value
+		// matching the filter value, so we can skip the first matcher
+		if len(matchers) > 1 {
+			matchers[0] = &anyMatcher{}
+		}
+
+		var iter indexIterator
+		if isUniqueFetchByFullKey(&f.indexDesc, fieldConditions) {
```

Comment:
todo: This looks like it will always be false no? As to get here at least one of the `fieldContitions` must be `opIn`, and `isUniqueFetchByFullKey` requires that all `fieldContitions` must be `opEq`.

Regardless of whether I'm right or wrong here, can you please document this.

Reply:
added a comment to `isUniqueFetchByFullKey`

Reply:
Ah of course, cheers - that comment really helps, I didn't spot it skipping the first index :)